### DISCUSSION TITLE
Rendering Problem - Canvas in React

### DISCUSSION THREAD

**Discussion by chris-kuehn** [neuefische/questions#675] [Web] [answered]
Hi,

I am using the Canvas API in my React app to draw an image. 
With the useRef() I create a reference to the canvas element, furthermore I use useEffect().
Unfortunately, my canvas is rendering incorrectly.

![Bildschirm­foto 2023-08-17 um 15 55 47](https://github.com/neuefische/questions/assets/136816603/21513f4f-675a-4e6c-a173-de1a18502685)

I ran the same statements I used for drawing into a vanilla JavaScrit environment. In this case, it rendered correctly.

![Bildschirm­foto 2023-08-17 um 15 55 57](https://github.com/neuefische/questions/assets/136816603/80bc5d46-7041-40aa-a8d0-55b3469274da)

My project is on GitHub and I have created a pull request.

[https://github.com/chris-kuehn/capstone-project/pull/4](https://github.com/chris-kuehn/capstone-project/pull/4)

About help I would be very happy! :)

Best regards

Christian

**Comment by Kheirah** [comment, marked as answer]
Hey Christian,

This weird behavior only appears in development mode.
React _runs every render twice_ when in development mode. That means, every component is run twice, every `useEffect` is run twice, and unfortunately your transformation on the canvas is run twice.

If you want to fix that while still developing on localhost, here's the missing piece to your puzzle:
```js
useEffect(() => {
    const canvas = canvasRef.current;
    const ctx = canvas.getContext("2d");
    // Begin drawing statements
    ctx.fillStyle = "hsl(113, 60%, 95%)";
    ctx.fillRect(0, 0, canvas.width, canvas.height);

    ctx.fillStyle = "hsl(148, 100%, 50%)";
    ctx.translate(150, 150);
    ctx.rotate((45 * Math.PI) / 180);
    ctx.fillRect(-50, -50, 100, 100);

    // Reset the transformation matrix to the identity matrix
    ctx.setTransform(1, 0, 0, 1, 0, 0);
    // End drawing statements
  }, []);
```
By resetting the transformation to its origin, the `useEffect` may runs twice, and three times, however many times, and all without distorting your rotated square. But like I said, this behavior does not appear in production. You can actually check your recent deployment and see for yourself.

By the way, this discussion board is kept in English. Try it next time 😄

**Reply by chris-kuehn** [reply]
Hi Keihra,

I have translated my question again in English so that all students can follow the problem. 

Thanks for your quick answer and explanation! 🫶 I added the line code and now it works great. :)

Best regards

Christian 😀
